### PR TITLE
get and set cameras on stitcher

### PR DESCRIPTION
**Pull request by amirtu**
resolves https://github.com/opencv/opencv/issues/18459

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/stitching/include/opencv2/stitching.hpp [hunk 1]
```diff
@@ -318,7 +319,7 @@ class CV_EXPORTS_W Stitcher
     std::vector<detail::MatchesInfo> pairwise_matches_;
     std::vector<cv::UMat> seam_est_imgs_;
     std::vector<int> indices_;
-    std::vector<detail::CameraParams> cameras_;
```
Comment (asmorkalov) [inline modules/stitching/include/opencv2/stitching.hpp:321]: Please restore `detail` namespace.

File: modules/stitching/include/opencv2/stitching.hpp [hunk 2]
```diff
@@ -286,7 +286,8 @@ class CV_EXPORTS_W Stitcher
     CV_WRAP Status stitch(InputArrayOfArrays images, InputArrayOfArrays masks, OutputArray pano);
 
     std::vector<int> component() const { return indices_; }
-    std::vector<detail::CameraParams> cameras() const { return cameras_; }
+    CV_WRAP std::vector<CameraParams> getCameras() const { return cameras_; }
```
Comment (asmorkalov) [inline modules/stitching/include/opencv2/stitching.hpp:289]: please restore original method signature and remove `CV_WRAP` prefix. It cannot be handled by bindings generator automatically.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/stitching/src/stitcher.cpp [hunk 3]
```diff
@@ -541,6 +540,102 @@ Stitcher::Status Stitcher::estimateCameraParams()
     return OK;
 }
 
+Stitcher::Status Stitcher::setTransform(InputArrayOfArrays images, const std::vector<detail::CameraParams> &cameras) {
```
Comment (amirtu) [inline modules/stitching/src/stitcher.cpp:543]: Probably the following code is excessive, as the same result could be achieved by reusing an initialized (through estimateTransform) stitcher's composePanorama with other images (i.e. different from those passed to estimateTransform). Providing the setter to the homographies (cameras array) would be sufficient.

File: modules/stitching/test/test_another_noname.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "test_precomp.hpp"
+
+namespace opencv_test { namespace {
+
+    TEST(Stitching, AnotherNoname)
```
Comment (asmorkalov) [inline modules/stitching/test/test_another_noname.cpp:9]: Please name the file and test properly.

File: modules/stitching/src/stitcher.cpp [hunk 5]
```diff
@@ -542,7 +542,7 @@ Stitcher::Status Stitcher::estimateCameraParams()
 
 Stitcher::Status Stitcher::setTransform(InputArrayOfArrays images, const std::vector<detail::CameraParams> &cameras) {
     std::vector<int> indices;
-    for (int i = 0; i < (int)imgs_.size(); i++)
+    for (int i = 0; i < (int)images.total(); i++)
```
Comment (asmorkalov) [inline modules/stitching/src/stitcher.cpp:546]: Why?

File: modules/stitching/include/opencv2/stitching.hpp [hunk 6]
```diff
@@ -259,8 +259,20 @@ class CV_EXPORTS_W Stitcher
      */
     CV_WRAP Status estimateTransform(InputArrayOfArrays images, InputArrayOfArrays masks = noArray());
 
+    /** @overload */
     Status setTransform(InputArrayOfArrays images, const std::vector<detail::CameraParams> &cameras);
 
+    /** @brief These functions reuse image matches and estimated rotations of each camera if this information is already
+     available.
+
+    @note Use the functions only if you're aware of the stitching pipeline, otherwise use
```
Comment (asmorkalov) [inline modules/stitching/include/opencv2/stitching.hpp:268]: I propose to replace the note with reference to `cameras()` method.

File: modules/stitching/include/opencv2/stitching.hpp [hunk 7]
```diff
@@ -259,6 +259,24 @@ class CV_EXPORTS_W Stitcher
      */
     CV_WRAP Status estimateTransform(InputArrayOfArrays images, InputArrayOfArrays masks = noArray());
 
+    /** @overload */
```
Comment (alalek) [inline modules/stitching/include/opencv2/stitching.hpp:262]: Overloads should go after the main function.

http://pullrequest.opencv.org/buildbot/export/pr/19619/docs/d2/d8d/classcv_1_1Stitcher.html#a1bbb3bb8cc7763a0177a05b21c97f794

> It differs from the above function only in what argument(s) it accepts.

Comment (asmorkalov) [reply modules/stitching/include/opencv2/stitching.hpp]: Done.

**Review by asmorkalov — APPROVED** [review]

**Pull request merged**